gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcDownscopingInterceptor.java

```diff
@@ -0,0 +1,226 @@
+
+  public GoogleCloudStorageClientGrpcDownscopingInterceptor(
+      Function<List<AccessBoundary>, String> downscopedAccessToken) {
+    this.downscopingFunction = downscopedAccessToken;
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    logger.atFinest().log(String.format("interceptCall(): method=%s", method.getFullMethodName()));
+    return new SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+      private int flowControlRequests;
+      private String methodName;
+      private Metadata headers;
+      private Listener<RespT> responseListener;
+
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        this.responseListener = responseListener;
+        this.headers = headers;
+        this.methodName = method.getFullMethodName();
+
+        logger.atFinest().log("start(): method=%s", methodName);
+      }
+
+      @Override
+      public void request(int numMessages) {
+        if (headers != null) {
+          this.flowControlRequests += numMessages;
+        } else {
+          super.request(numMessages);
+        }
+      }
+
+      @Override
+      public void sendMessage(ReqT message) {
+        setAuthHeader(message);
+        if (headers != null) {
+          delegate().start(responseListener, headers);
+        }
+
+        if (flowControlRequests != 0) {
+          super.request(flowControlRequests);
+        }
+        headers = null;
+
+        super.sendMessage(message);
+      }
+
+      private void setAuthHeader(ReqT message) {
+        if (headers == null) {
+          return;
+        }
+
+        String token = getDownScopedToken(message);
+        if (token != null) {
+          logger.atFinest().log("Setting down-scoped auth token");
+          headers.put(AUTH_KEY, "Bearer " + token);
+        }
+      }
+
+      private String getDownScopedToken(ReqT message) {
+        List<AccessBoundary> accessBoundaries = getAccessBoundaries(message);
+        return getDownscopedToken(accessBoundaries);
+      }
+
+      private String getBucketName(String bucketLongName) {
+        return bucketLongName.substring(BUCKET_PREFIX_LENGTH);
+      }
+
+      private List<AccessBoundary> getAccessBoundaries(ReqT message) {
+        if (GOOGLE_STORAGE_V_2_STORAGE_READ_OBJECT.equals(methodName)) {
+          ReadObjectRequest readObjectRequest = (ReadObjectRequest) message;
+          String bucketName = getBucketName(readObjectRequest.getBucket());
+          String objectName = readObjectRequest.getObject();
+
+          return ImmutableList.of(
+              AccessBoundary.create(bucketName, objectName, Action.READ_OBJECTS));
+        }
+
+        if (GOOGLE_STORAGE_V_2_STORAGE_WRITE_OBJECT.equals(methodName)) {
+          WriteObjectSpec writeObjectSpec = ((WriteObjectRequest) message).getWriteObjectSpec();
```

Comment:
for resumable upload specific write requests Resource info will not be there any only uploadId would be present.

Do we have integration test for it?

Reply:
Test this out. I am not seeing the behavior you are describing. We do not have integration tests for Downscoped tokens.

Reply:
[`WriteObjectRequest`](https://github.com/googleapis/java-storage/blob/v2.22.1/proto-google-cloud-storage-v2/src/main/java/com/google/storage/v2/WriteObjectRequest.java#L71) captures what writeObjectSpec will capturing i.e. either uploadId or resource

Reply:
Discussed offline.